test/vcf/report_writer_test.cpp

```diff
@@ -140,63 +141,141 @@ namespace ebi
+          errorDAO.for_each_error([&](std::shared_ptr<ebi::vcf::Error> error) {
+              CHECK(error->get_line() == line);
+              CHECK(error->get_raw_message() == message);
+              errors_read++;
+          });
+          CHECK(errors_read == 1);
+      }
+      
+      SECTION("write and read warnings")
+      {
+          size_t line = 10;
+          std::string message{"testing warnings"};
+          ebi::vcf::Error test_error{line, message};
+          errorDAO.write_warning(test_error);
+          errorDAO.flush();
+          
+          size_t errors_read = 0;
+          errorDAO.for_each_warning([&](std::shared_ptr<ebi::vcf::Error> error) {
+              CHECK(error->get_line() == line);
+              CHECK(error->get_raw_message() == message);
+              errors_read++;
+          });
+          CHECK(errors_read == 1);
+      }
+      
+      SECTION("write and read error codes")
+      {
+          size_t line = 8;
+          std::string message{"testing erros"};
+          ebi::vcf::Error generic_error{line, message};
+          ebi::vcf::MetaSectionError meta_section_error{line, message};
+          ebi::vcf::SamplesBodyError samples_body_error{line, message};
+          errorDAO.write_error(generic_error);
+          errorDAO.write_error(meta_section_error);
+          errorDAO.write_error(samples_body_error);
+          errorDAO.flush();
+          
+          std::vector<std::shared_ptr<ebi::vcf::Error>> errors;
+          errorDAO.for_each_error([&](std::shared_ptr<ebi::vcf::Error> error) {
+              errors.push_back(error);
+          });
+          
+          CHECK(errors.size());
+          CHECK(errors[0]->get_code() == ebi::vcf::ErrorCode::error);
+          CHECK(errors[1]->get_code() == ebi::vcf::ErrorCode::meta_section);
+          CHECK(errors[2]->get_code() == ebi::vcf::ErrorCode::samples_body);
+      }
+      
       boost::filesystem::path db_file{db_name};
       boost::filesystem::remove(db_file);
+      CHECK_FALSE(boost::filesystem::exists(db_file));
   }
 
   TEST_CASE("integration test: validator and sqlite", "[output]")
   {
       auto path = boost::filesystem::path("test/input_files/failed/failed_fileformat_000.vcf");
 
       std::string db_name = path.string() + ".errors.db";
-      sqlite3* db;
 
-      SECTION(path.string())
       {
-          {
-              std::shared_ptr<ebi::vcf::ReportWriter> output{std::make_shared<ebi::vcf::SqliteReportWriter>(db_name)};
-              CHECK_FALSE(is_valid(path.string(), *output));
-          }
-          int rc = sqlite3_open(db_name.c_str(), &db);
-          if(rc != SQLITE_OK) {
-              sqlite3_close(db);
-              throw std::runtime_error(std::string("Can't open database: ") + sqlite3_errmsg(db));
-          }
+          std::shared_ptr<ebi::vcf::ReportWriter> output{std::make_shared<ebi::vcf::SqliteReportRW>(db_name)};
+          CHECK_FALSE(is_valid(path.string(), *output));
+      }
```

Comment:
The extra block is needed to flush, SqliteReportRW provides a RAII-style for flushing the writings. It provides `close()` and `flush()` as well, but we are handling the interface ReportWriter here. To remove the block we can add `close()` and `flush()` methods to the ReportWriter interface.